File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/ActivityModels.kt
package tech.mmarca.openvitals.domain.model

import java.time.Instant
import java.time.ZoneOffset
import java.time.LocalDate

data class ExerciseData(
    val id: String,
    val title: String?,
    val exerciseType: Int,
    val startTime: Instant,
    val endTime: Instant,
    val durationMs: Long,
    val source: String,
    val totalDistanceMeters: Double? = null,
    val totalCaloriesKcal: Double? = null,
    val activeCaloriesKcal: Double? = null,
    val steps: Long? = null,
    val wheelchairPushes: Long? = null,
    val averageSpeedMetersPerSecond: Double? = null,
    val averagePowerWatts: Double? = null,
    val averageStepsCadenceRate: Double? = null,
    val averageCyclingCadenceRpm: Double? = null,
    val averageHeartRateBpm: Long? = null,
    val floorsClimbed: Int? = null,
    val elevationGainedMeters: Double? = null,
    val notes: String? = null,
    val startZoneOffset: ZoneOffset? = null,
    val endZoneOffset: ZoneOffset? = null,
    val lastModifiedTime: Instant? = null,
    val clientRecordId: String? = null,
    val clientRecordVersion: Long? = null,
    val recordingMethod: Int? = null,
    val device: ExerciseDeviceData? = null,
    val plannedExerciseSessionId: String? = null,
    val segments: List<ExerciseSegmentData> = emptyList(),
    val laps: List<ExerciseLapData> = emptyList(),
    val route: ExerciseRouteData = ExerciseRouteData(),
    val isOpenVitalsEntry: Boolean = false,
    val totalCaloriesSource: CaloriesBurnedSource = if (totalCaloriesKcal != null) {
        CaloriesBurnedSource.RECORDED_TOTAL
    } else {
        CaloriesBurnedSource.NO_DATA
    },
) {
    val durationMinutes: Long get() = durationMs / 60_000
}

data class ExerciseDeviceData(
    val type: Int,
    val manufacturer: String?,
    val model: String?,
)

data class ExerciseSegmentData(
    val startTime: Instant,
    val endTime: Instant,
    val segmentType: Int,
    val repetitions: Int,
    val setIndex: Int? = null,
    /** Load lifted in this set, when recorded. */
    val weightKg: Double? = null,
) {
    val durationMs: Long get() = endTime.toEpochMilli() - startTime.toEpochMilli()
}

data class ExerciseLapData(
    val startTime: Instant,
    val endTime: Instant,
    val lengthMeters: Double?,
) {
    val durationMs: Long get() = endTime.toEpochMilli() - startTime.toEpochMilli()
}

data class ActivityRecordingLap(
    val startTime: Instant,
    val endTime: Instant,
    val distanceMeters: Double?,
) {
    val durationMs: Long get() = endTime.toEpochMilli() - startTime.toEpochMilli()
}

data class ActivityRecordingMarker(
    val id: String,
    val time: Instant,
    val latitude: Double,
    val longitude: Double,
    val altitudeMeters: Double?,
    val name: String,
    val note: String = "",
    val type: String = ActivityRecordingMarkerType.Generic.value,
)

enum class ActivityRecordingMarkerType(val value: String) {
    Generic("generic"),
}

data class ExerciseRouteData(
    val status: ExerciseRouteStatus = ExerciseRouteStatus.NO_DATA,
    val points: List<ExerciseRoutePoint> = emptyList(),
)

enum class ExerciseRouteStatus {
    DATA,
    CONSENT_REQUIRED,
    NO_DATA,
}

data class ExerciseRoutePoint(
    val time: Instant,
    val latitude: Double,
    val longitude: Double,
    val altitudeMeters: Double?,
    val horizontalAccuracyMeters: Double?,
    val verticalAccuracyMeters: Double?,
)

data class ActivityPauseInterval(
    val startTime: Instant,
    val endTime: Instant,
)

data class ActivityExerciseSegmentWrite(
    val startTime: Instant,
    val endTime: Instant,
    val segmentType: Int,
    val repetitions: Int = 0,
    val setIndex: Int? = null,
    val weightKg: Double? = null,
)

/** Where an activity's data came from. Decides the recording method Health Connect stores. */
enum class ActivityRecordSource {
    /** Typed in, or recorded by this app on the phone. */
    MANUAL_ENTRY,

    /** A file a paired watch recorded. */
    WATCH,

    /** A file the user picked. Some device recorded it; which one is unknown. */
    FILE,
}

/** A total the activity form owns. Each is stored as one record beside the session. */
enum class ActivityFormMetric { STEPS, DISTANCE, ELEVATION, ACTIVE_CALORIES, TOTAL_CALORIES }

/**
 * What OpenVitals itself stored for one session. Null means it stored nothing for that
 * total, whatever other apps recorded over the same minutes.
 */
data class OwnActivityMetrics(
    val steps: Long? = null,
    val distanceMeters: Double? = null,
    val elevationGainedMeters: Double? = null,
    val activeCaloriesKcal: Double? = null,
    val totalCaloriesKcal: Double? = null,
)

data class ActivityWriteRequest(
    val exerciseType: Int,
    val startTime: Instant,
    val endTime: Instant,
    val title: String? = null,
    val notes: String? = null,
    val plannedExerciseSessionId: String? = null,
    val routePoints: List<ExerciseRoutePoint> = emptyList(),
    val pauseIntervals: List<ActivityPauseInterval> = emptyList(),
    val laps: List<ExerciseLapData> = emptyList(),
    val exerciseSegments: List<ActivityExerciseSegmentWrite> = emptyList(),
    val stepsCount: Long? = null,
    val distanceMeters: Double? = null,
    val elevationGainedMeters: Double? = null,
    val activeCaloriesKcal: Double? = null,
    val totalCaloriesKcal: Double? = null,
    val bleSamples: BleRecordingSampleBuffer = BleRecordingSampleBuffer(),
    val source: ActivityRecordSource = ActivityRecordSource.MANUAL_ENTRY,
    /**
     * A stable key for an imported file. With it every client record id is a function of the
     * file, so importing the file again updates the records instead of adding a second set.
     */
    val importKey: String? = null,
    /**
     * For an edit: the totals the user changed. Every other total keeps the value OpenVitals
     * stored, including one the form does not show for this activity type. Null means every
     * total in this request is a decision, as for a new entry.
     */
    val editedMetrics: Set<ActivityFormMetric>? = null,
) {
    /** This request with every total the user did not change taken from [own]. */
    fun keepingUntouchedMetrics(own: OwnActivityMetrics): ActivityWriteRequest {
        val edited = editedMetrics ?: return this
        return copy(
            stepsCount = if (ActivityFormMetric.STEPS in edited) stepsCount else own.steps,
            distanceMeters = if (ActivityFormMetric.DISTANCE in edited) distanceMeters else own.distanceMeters,
            elevationGainedMeters =
            if (ActivityFormMetric.ELEVATION in edited) elevationGainedMeters else own.elevationGainedMeters,
            activeCaloriesKcal =
            if (ActivityFormMetric.ACTIVE_CALORIES in edited) activeCaloriesKcal else own.activeCaloriesKcal,
            totalCaloriesKcal =
            if (ActivityFormMetric.TOTAL_CALORIES in edited) totalCaloriesKcal else own.totalCaloriesKcal,
        )
    }
}

data class PlannedExerciseData(
    val id: String,
    val title: String?,
    val exerciseType: Int,
    val startTime: Instant,
    val endTime: Instant,
    /** The offset the plan was written in; the day it belongs to is read in this, not the phone's current zone. */
    val startZoneOffset: ZoneOffset? = null,
    val hasExplicitTime: Boolean,
    val completedExerciseSessionId: String?,
    val notes: String?,
    val blockCount: Int,
    val source: String,
    /**
     * The raw `dataOrigin` package. [source] shows a synced record's original
     * app, so ownership gates must use this.
     */
    val dataOriginPackage: String = source,
    val blocks: List<PlannedExerciseBlockData> = emptyList(),
) {
    val durationMs: Long get() = endTime.toEpochMilli() - startTime.toEpochMilli()
}

data class PlannedExerciseBlockData(
    val repetitions: Int,
    val description: String?,
    val steps: List<PlannedExerciseStepData>,
)

data class PlannedExerciseStepData(
    val exerciseType: Int,
    val exercisePhase: Int,
    val description: String?,
    val completion: PlannedExerciseCompletion,
    /** Pace/power/heart-rate/… targets other apps attach; carried through untouched. */
    val performanceTargets: List<PlannedExercisePerformanceTarget> = emptyList(),
)

sealed interface PlannedExerciseCompletion {
    data class Repetitions(val repetitions: Int) : PlannedExerciseCompletion
    data class DurationSeconds(val seconds: Long) : PlannedExerciseCompletion
    data class DistanceMeters(val meters: Double) : PlannedExerciseCompletion
    data class DistanceAndDuration(val meters: Double, val seconds: Long) : PlannedExerciseCompletion
    data class Steps(val steps: Int) : PlannedExerciseCompletion
    data class ActiveCaloriesKcal(val kcal: Double) : PlannedExerciseCompletion
    data class TotalCaloriesKcal(val kcal: Double) : PlannedExerciseCompletion
    data object Manual : PlannedExerciseCompletion
    data object Unknown : PlannedExerciseCompletion
}

sealed interface PlannedExercisePerformanceTarget {
    data class Power(val minWatts: Double, val maxWatts: Double) : PlannedExercisePerformanceTarget
    data class Speed(val minMetersPerSecond: Double, val maxMetersPerSecond: Double) : PlannedExercisePerformanceTarget
    data class Cadence(val minRpm: Double, val maxRpm: Double) : PlannedExercisePerformanceTarget
    data class HeartRate(val minBpm: Double, val maxBpm: Double) : PlannedExercisePerformanceTarget
    data class Weight(val kilograms: Double) : PlannedExercisePerformanceTarget
    data class RateOfPerceivedExertion(val rpe: Int) : PlannedExercisePerformanceTarget
    data object Amrap : PlannedExercisePerformanceTarget
    data object Unknown : PlannedExercisePerformanceTarget
}

data class PlannedExerciseWriteRequest(
    val id: String? = null,
    val exerciseType: Int,
    val startTime: Instant,
    val endTime: Instant,
    val title: String? = null,
    val notes: String? = null,
    val blocks: List<PlannedExerciseBlockData>,
)

data class DailySteps(
    val date: LocalDate,
    val steps: Long,
    val distanceMeters: Double,
    val wheelchairPushes: Long? = null,
    val floorsClimbed: Int? = null,
    val activeCaloriesKcal: Double? = null,
    val elevationGainedMeters: Double? = null,
)

data class ActivityProgressPoint(
    val time: Instant,
    val totalSteps: Long,
    val totalDistanceMeters: Double?,
    val totalCaloriesBurnedKcal: Double?,
    val totalActiveCaloriesKcal: Double? = null,
    val totalWheelchairPushes: Long? = null,
    val totalFloorsClimbed: Int? = null,
    val totalElevationGainedMeters: Double? = null,
)

data class SpeedSample(
    val time: Instant,
    val metersPerSecond: Double,
    val source: String,
)

enum class ActivityCadenceKind {
    CYCLING,
    STEPS,
}

data class ActivityCadenceSample(
    val time: Instant,
    val rate: Double,
    val kind: ActivityCadenceKind,
    val source: String,
)

File: lib/domain/model/activity_models.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import '../../core/time/local_date.dart';
import 'ble_sensor_models.dart';
import 'nutrition_models.dart';

part 'activity_models.freezed.dart';

@freezed
abstract class ExerciseData with _$ExerciseData {
  const ExerciseData._();

  const factory ExerciseData.build({
    required String id,
    required String? title,
    required int exerciseType,
    required DateTime startTime,
    required DateTime endTime,
    required int durationMs,
    required String source,
    required double? totalDistanceMeters,
    required double? totalCaloriesKcal,
    required double? activeCaloriesKcal,
    required int? steps,
    required int? wheelchairPushes,
    required double? averageSpeedMetersPerSecond,
    required double? averagePowerWatts,
    required double? averageStepsCadenceRate,
    required double? averageCyclingCadenceRpm,
    required int? averageHeartRateBpm,
    required int? floorsClimbed,
    required double? elevationGainedMeters,
    required String? notes,
    required Duration? startZoneOffset,
    required Duration? endZoneOffset,
    required DateTime? lastModifiedTime,
    required String? clientRecordId,
    required int? clientRecordVersion,
    required int? recordingMethod,
    required ExerciseDeviceData? device,
    required String? plannedExerciseSessionId,
    required List<ExerciseSegmentData> segments,
    required List<ExerciseLapData> laps,
    required ExerciseRouteData route,
    required bool isOpenVitalsEntry,
    required CaloriesBurnedSource totalCaloriesSource,
  }) = _ExerciseData;

  factory ExerciseData({
    required String id,
    required String? title,
    required int exerciseType,
    required DateTime startTime,
    required DateTime endTime,
    required int durationMs,
    required String source,
    double? totalDistanceMeters,
    double? totalCaloriesKcal,
    double? activeCaloriesKcal,
    int? steps,
    int? wheelchairPushes,
    double? averageSpeedMetersPerSecond,
    double? averagePowerWatts,
    double? averageStepsCadenceRate,
    double? averageCyclingCadenceRpm,
    int? averageHeartRateBpm,
    int? floorsClimbed,
    double? elevationGainedMeters,
    String? notes,
    Duration? startZoneOffset,
    Duration? endZoneOffset,
    DateTime? lastModifiedTime,
    String? clientRecordId,
    int? clientRecordVersion,
    int? recordingMethod,
    ExerciseDeviceData? device,
    String? plannedExerciseSessionId,
    List<ExerciseSegmentData> segments = const <ExerciseSegmentData>[],
    List<ExerciseLapData> laps = const <ExerciseLapData>[],
    ExerciseRouteData route = const ExerciseRouteData(),
    bool isOpenVitalsEntry = false,
    CaloriesBurnedSource? totalCaloriesSource,
  }) =>
      ExerciseData.build(
        id: id,
        title: title,
        exerciseType: exerciseType,
        startTime: startTime,
        endTime: endTime,
        durationMs: durationMs,
        source: source,
        totalDistanceMeters: totalDistanceMeters,
        totalCaloriesKcal: totalCaloriesKcal,
        activeCaloriesKcal: activeCaloriesKcal,
        steps: steps,
        wheelchairPushes: wheelchairPushes,
        averageSpeedMetersPerSecond: averageSpeedMetersPerSecond,
        averagePowerWatts: averagePowerWatts,
        averageStepsCadenceRate: averageStepsCadenceRate,
        averageCyclingCadenceRpm: averageCyclingCadenceRpm,
        averageHeartRateBpm: averageHeartRateBpm,
        floorsClimbed: floorsClimbed,
        elevationGainedMeters: elevationGainedMeters,
        notes: notes,
        startZoneOffset: startZoneOffset,
        endZoneOffset: endZoneOffset,
        lastModifiedTime: lastModifiedTime,
        clientRecordId: clientRecordId,
        clientRecordVersion: clientRecordVersion,
        recordingMethod: recordingMethod,
        device: device,
        plannedExerciseSessionId: plannedExerciseSessionId,
        segments: segments,
        laps: laps,
        route: route,
        isOpenVitalsEntry: isOpenVitalsEntry,
        totalCaloriesSource: totalCaloriesSource ??
            (totalCaloriesKcal != null
                ? CaloriesBurnedSource.recordedTotal
                : CaloriesBurnedSource.noData),
      );

  int get durationMinutes => durationMs ~/ 60000;
}

@freezed
abstract class ExerciseDeviceData with _$ExerciseDeviceData {
  const factory ExerciseDeviceData({
    required int type,
    required String? manufacturer,
    required String? model,
  }) = _ExerciseDeviceData;
}

@freezed
abstract class ExerciseSegmentData with _$ExerciseSegmentData {
  const ExerciseSegmentData._();

  const factory ExerciseSegmentData({
    required DateTime startTime,
    required DateTime endTime,
    required int segmentType,
    required int repetitions,
    int? setIndex,
  }) = _ExerciseSegmentData;

  int get durationMs =>
      endTime.millisecondsSinceEpoch - startTime.millisecondsSinceEpoch;
}

@freezed
abstract class ExerciseLapData with _$ExerciseLapData {
  const ExerciseLapData._();

  const factory ExerciseLapData({
    required DateTime startTime,
    required DateTime endTime,
    required double? lengthMeters,
  }) = _ExerciseLapData;

  int get durationMs =>
      endTime.millisecondsSinceEpoch - startTime.millisecondsSinceEpoch;
}

@freezed
abstract class ActivityRecordingLap with _$ActivityRecordingLap {
  const ActivityRecordingLap._();

  const factory ActivityRecordingLap({
    required DateTime startTime,
    required DateTime endTime,
    required double? distanceMeters,
  }) = _ActivityRecordingLap;

  int get durationMs =>
      endTime.millisecondsSinceEpoch - startTime.millisecondsSinceEpoch;
}

@freezed
abstract class ActivityRecordingMarker with _$ActivityRecordingMarker {
  const factory ActivityRecordingMarker({
    required String id,
    required DateTime time,
    required double latitude,
    required double longitude,
    required double? altitudeMeters,
    required String name,
    @Default('') String note,
    // Equals ActivityRecordingMarkerType.generic.value.
    @Default('generic') String type,
  }) = _ActivityRecordingMarker;
}

enum ActivityRecordingMarkerType {
  generic('generic');

  const ActivityRecordingMarkerType(this.value);

  final String value;
}

@freezed
abstract class ExerciseRouteData with _$ExerciseRouteData {
  const factory ExerciseRouteData({
    @Default(ExerciseRouteStatus.noData) ExerciseRouteStatus status,
    @Default(<ExerciseRoutePoint>[]) List<ExerciseRoutePoint> points,
  }) = _ExerciseRouteData;
}

enum ExerciseRouteStatus {
  data('DATA'),
  consentRequired('CONSENT_REQUIRED'),
  noData('NO_DATA');

  const ExerciseRouteStatus(this.storageName);

  /// Original Kotlin `.name` used for persistence round-trips.
  final String storageName;

  static ExerciseRouteStatus? fromStorage(String value) {
    for (final entry in values) {
      if (entry.storageName == value) return entry;
    }
    return null;
  }
}

@freezed
abstract class ExerciseRoutePoint with _$ExerciseRoutePoint {
  const factory ExerciseRoutePoint({
    required DateTime time,
    required double latitude,
    required double longitude,
    required double? altitudeMeters,
    required double? horizontalAccuracyMeters,
    required double? verticalAccuracyMeters,
  }) = _ExerciseRoutePoint;
}

@freezed
abstract class ActivityPauseInterval with _$ActivityPauseInterval {
  const factory ActivityPauseInterval({
    required DateTime startTime,
    required DateTime endTime,
  }) = _ActivityPauseInterval;
}

@freezed
abstract class ActivityExerciseSegmentWrite with _$ActivityExerciseSegmentWrite {
  const factory ActivityExerciseSegmentWrite({
    required DateTime startTime,
    required DateTime endTime,
    required int segmentType,
    @Default(0) int repetitions,
    int? setIndex,
  }) = _ActivityExerciseSegmentWrite;
}

@freezed
abstract class ActivityWriteRequest with _$ActivityWriteRequest {
  const factory ActivityWriteRequest({
    required int exerciseType,
    required DateTime startTime,
    required DateTime endTime,
    String? title,
    String? notes,
    String? plannedExerciseSessionId,
    @Default(<ExerciseRoutePoint>[]) List<ExerciseRoutePoint> routePoints,
    @Default(<ActivityPauseInterval>[])
    List<ActivityPauseInterval> pauseIntervals,
    @Default(<ExerciseLapData>[]) List<ExerciseLapData> laps,
    @Default(<ActivityExerciseSegmentWrite>[])
    List<ActivityExerciseSegmentWrite> exerciseSegments,
    int? stepsCount,
    double? distanceMeters,
    double? elevationGainedMeters,
    double? activeCaloriesKcal,
    double? totalCaloriesKcal,
    @Default(BleRecordingSampleBuffer()) BleRecordingSampleBuffer bleSamples,
  }) = _ActivityWriteRequest;
}

@freezed
abstract class PlannedExerciseData with _$PlannedExerciseData {
  const PlannedExerciseData._();

  const factory PlannedExerciseData({
    required String id,
    required String? title,
    required int exerciseType,
    required DateTime startTime,
    required DateTime endTime,
    required bool hasExplicitTime,
    required String? completedExerciseSessionId,
    required String? notes,
    required int blockCount,
    required String source,
    @Default(<PlannedExerciseBlockData>[])
    List<PlannedExerciseBlockData> blocks,
  }) = _PlannedExerciseData;

  int get durationMs =>
      endTime.millisecondsSinceEpoch - startTime.millisecondsSinceEpoch;
}

@freezed
abstract class PlannedExerciseBlockData with _$PlannedExerciseBlockData {
  const factory PlannedExerciseBlockData({
    required int repetitions,
    required String? description,
    required List<PlannedExerciseStepData> steps,
  }) = _PlannedExerciseBlockData;
}

@freezed
abstract class PlannedExerciseStepData with _$PlannedExerciseStepData {
  const factory PlannedExerciseStepData({
    required int exerciseType,
    required int exercisePhase,
    required String? description,
    required PlannedExerciseCompletion completion,
  }) = _PlannedExerciseStepData;
}

/// Sealed hierarchy mirroring Kotlin `sealed interface
/// PlannedExerciseCompletion`. Nested Kotlin members are flattened with a
/// `PlannedExerciseCompletion` prefix.
sealed class PlannedExerciseCompletion {
  const PlannedExerciseCompletion();
}

class PlannedExerciseCompletionRepetitions extends PlannedExerciseCompletion {
  const PlannedExerciseCompletionRepetitions(this.repetitions);

  final int repetitions;

  @override
  bool operator ==(Object other) =>
      other is PlannedExerciseCompletionRepetitions &&
      other.repetitions == repetitions;

  @override
  int get hashCode => repetitions.hashCode;
}

class PlannedExerciseCompletionDurationSeconds
    extends PlannedExerciseCompletion {
  const PlannedExerciseCompletionDurationSeconds(this.seconds);

  final int seconds;

  @override
  bool operator ==(Object other) =>
      other is PlannedExerciseCompletionDurationSeconds &&
      other.seconds == seconds;

  @override
  int get hashCode => seconds.hashCode;
}

class PlannedExerciseCompletionManual extends PlannedExerciseCompletion {
  const PlannedExerciseCompletionManual();
}

class PlannedExerciseCompletionUnknown extends PlannedExerciseCompletion {
  const PlannedExerciseCompletionUnknown();
}

@freezed
abstract class PlannedExerciseWriteRequest with _$PlannedExerciseWriteRequest {
  const factory PlannedExerciseWriteRequest({
    String? id,
    required int exerciseType,
    required DateTime startTime,
    required DateTime endTime,
    String? title,
    String? notes,
    required List<PlannedExerciseBlockData> blocks,
  }) = _PlannedExerciseWriteRequest;
}

@freezed
abstract class DailySteps with _$DailySteps {
  const factory DailySteps({
    required LocalDate date,
    required int steps,
    required double distanceMeters,
    int? wheelchairPushes,
    int? floorsClimbed,
    double? activeCaloriesKcal,
    double? elevationGainedMeters,
  }) = _DailySteps;
}

@freezed
abstract class ActivityProgressPoint with _$ActivityProgressPoint {
  const factory ActivityProgressPoint({
    required DateTime time,
    required int totalSteps,
    required double? totalDistanceMeters,
    required double? totalCaloriesBurnedKcal,
    double? totalActiveCaloriesKcal,
    int? totalWheelchairPushes,
    int? totalFloorsClimbed,
    double? totalElevationGainedMeters,
  }) = _ActivityProgressPoint;
}

@freezed
abstract class SpeedSample with _$SpeedSample {
  const factory SpeedSample({
    required DateTime time,
    required double metersPerSecond,
    required String source,
  }) = _SpeedSample;
}

enum ActivityCadenceKind {
  cycling('CYCLING'),
  steps('STEPS');

  const ActivityCadenceKind(this.storageName);

  /// Original Kotlin `.name` used for persistence round-trips.
  final String storageName;

  static ActivityCadenceKind? fromStorage(String value) {
    for (final entry in values) {
      if (entry.storageName == value) return entry;
    }
    return null;
  }
}

@freezed
abstract class ActivityCadenceSample with _$ActivityCadenceSample {
  const factory ActivityCadenceSample({
    required DateTime time,
    required double rate,
    required ActivityCadenceKind kind,
    required String source,
  }) = _ActivityCadenceSample;
}
